refactor: move graphql client to context

sethidden opened the pull request:
this is so you can query graphql endpoints in plugins

The getHeaderProxy(ctx) may seem a little convoluted but:
- graphql-request doesn't support dynamic headers https://github.com/prisma-labs/graphql-request/pull/91
- it's possible to just dynamically calculate getHeaders() before each request in useApi (just add a third argument to "request"), but that way if someone wants to query something from the a Nuxt plugin they'd need to reimplement the cookie logic

This PR also makes sure that the clients are instantiated only once - before each new request had a new client

atagisow-alokai committed:
refactor: remove cookie instead of empty set

test: fix tests

sethidden commented on packages/theme/plugins/graphqlClient.ts:
```diff
@@ -0,0 +1,84 @@
+type GraphQlClients = Record<'query' | 'mutation', GraphQLClient>;
+
+const plugin : Plugin = (ctx, inject) => {
+  // TODO remember to remove the magento http options aswell
```
this

sethidden commented:
@Frodigo do you still want me to mark the useApi composable as deprecated? I'd probably need to make a task to replace all the ocurrences

sethidden commented on packages/theme/composables/useApi/index.ts:
```diff
@@ -1,54 +1,14 @@
-  const magentoConfig = app.$vsf.$magento.config;
-  // TODO remove once we remove apollo client
-  const { useGETForQueries } = magentoConfig.customApolloHttpLinkOptions;
-  const defaultEndpoint = magentoConfig.magentoApiEndpoint;
```
Hmm magentoApiEndpoint also seems like a candidate to be removed completely from the codebase (we have env var)

Frodigo commented on docs/guide/graphql-get.md:
```diff
@@ -45,10 +45,7 @@ sub process_graphql_headers {
 
-3. Open the `middleware.config.js` file, it should be located on the root folder of your store, there you will change the flag `useGETForQueries` to `true` to enable the usage of GET request for queries
-
+3. Make sure your GraphQL client sends queries (but not mutations) using the HTTP GET method. Configuring this for `graphql-request` is as follows:
```
I think we shouldn't remove that point 3 because it's still valid and it's about server middleware. 

There should be another document that describes the new GraphQL, and actually, we have a ticket in the backlog for that, so let's restore this change, please.

Frodigo replied:
Som in the new API client we won't use that `useGetForQueries` config option, but documentation will be done in separate PR.

Frodigo commented on packages/api-client/src/types/setup.ts:
```diff
@@ -92,7 +91,6 @@ export interface Config<T = any> extends ClientConfig {
   client?: ApolloClient<T>;
   storage: Storage;
   customOptions?: ApolloClientOptions<any>;
-  customApolloHttpLinkOptions?: HttpOptions;
```
Related to the previous comment, please restore this change.

Frodigo commented on packages/api-client/src/helpers/magentoLink/graphQl.ts:
```diff
@@ -67,7 +67,6 @@ export const apolloLinkFactory = (settings: Config, handlers?: {
     fetchOptions: {
       agent,
     },
-    ...settings.customApolloHttpLinkOptions,
```
Related to the previous comment, please restore this change.

Frodigo commented on packages/theme/middleware.config.js:
```diff
@@ -21,9 +21,6 @@ module.exports = {
             default: isCheckoutEnabled,
           },
         },
-        customApolloHttpLinkOptions: {
```
As I mentioned before we won't remove this.

Frodigo commented on packages/theme/plugins/graphqlClient.ts:
```diff
@@ -0,0 +1,84 @@
+import { GraphQLClient } from 'graphql-request';
+import { Plugin, Context } from '@nuxt/types';
```
```suggestion
import type { Plugin, Context } from '@nuxt/types';
```

Frodigo commented on packages/theme/plugins/graphqlClient.ts:
```diff
@@ -0,0 +1,84 @@
+  return new Proxy({}, proxyHandler);
+};
+
+type GraphQlClients = Record<'query' | 'mutation', GraphQLClient>;
```
The name of this is weird to me. WHat do you think about `GraphQlOperations`?

sethidden replied:
Well, they aren't operations. They're instances of a client that can make requests. This is a different approach from what useApi used which created a new client for each request. So the approach here is "upside down" as to what was with useApi

sethidden replied:
i.e. you still need to do ctx.$graphql.query.request(), it's not query()

Frodigo replied:
Ok, I see, thanks, I understand now so no changes are required.

Frodigo commented on packages/theme/nuxt.config.js:
```diff
@@ -226,6 +225,9 @@ export default () => {
     image: {
       provider: process.env.VSF_IMAGE_PROVIDER,
     },
+    env: {
```
Why this is needed here? I  see in the code that you use `process.env.VSF_MAGENTO_GRAPHQL_URL;` so I don't understand the purpose of this config.

sethidden replied:
Environment variables that are passed to nuxt are only available server side by default. So if I tried to go somewhere client-side the requests would fail because their VSF_MAGENTO_GRAPHQL_URL would be undefined.

This works without useApi beacuse it uses the middleware.config.js magentoApiEndpoint https://github.com/vuestorefront/magento2/pull/839/files#diff-66e4ffc399d4ad1f1464d30529bbb1ff498032d4293f1b3e83fdb33d4ee42e6dL12

Since you said you prefer not to modify api-config things (so I can't remove magentoApiEndpoint) I could either:
1. remove https://github.com/vuestorefront/magento2/pull/839/files#diff-ceda593cdb4e231ccc447030a0fd70207f4b6d2f7cbd5e6cff74d1309270c5e7R62 and keep using magentoApiEndpoint
2. keep using VSF_MAGENTO_GRAPHQL_URL directly since it's one layer less

what do you think?

Frodigo replied:
I see, let's go with point 2. keep using VSF_MAGENTO_GRAPHQL_URL directly since it's one layer less

atagisow-alokai committed:
refactor: move graphqlclient from useApi to context

M2-354

this is so you can query graphql endpoints in plugins

refactor: useApi to use client from context

refactor: move dynamic headers to client instance

I don't want to keep the append the headers in useApi because if someone
wants to pull the graphql clients from the context - e.g. when
requesting something in a plugin - they'll have to reimplement the
cookie logic again

this commit is a checkpoint commit - the '' fallback in graphqlClient
isn't ideal because the header shouldn't be appended to the request if
it's empty. Getters are limited int this way. I'll try something with Proxy

refactor: proxy workaround for dynamic headers

refactor: use proxies as workaround for dynamic headers

refactor: add template types to useApi

refactor: remove references to customApolloHttpLinkOptions

fix: fix spread

refactor: code review changes
